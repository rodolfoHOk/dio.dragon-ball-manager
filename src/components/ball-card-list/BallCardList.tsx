import { useEffect, useState } from 'react';
import { Ball } from '../../models/ball';
import { BallFilterForm } from './BallFilterForm';
import { BallCard } from './BallCard';

type BallCardListProps = {
  profileId: number;
  allBalls: Ball[];
};

export function BallCardList({ profileId, allBalls }: BallCardListProps) {
  const [selected, setSelected] = useState('all');
  const [balls, setBalls] = useState<Ball[]>([]);

  function onSelect(value: string) {
    setSelected(value);
  }

  useEffect(() => {
    console.log(selected);
    if (selected === 'me') {
      const balls = allBalls.filter((ball) => ball.owner === profileId);
      setBalls(balls);
    } else if (selected === 'not-me') {
      const balls = allBalls.filter((ball) => ball.owner !== profileId);
      setBalls(balls);
    } else {
      setBalls(allBalls);
    }
  }, [selected, allBalls, profileId]);

  return (
    <>
      <div className="flex justify-between items-center">
        <h1 className="font-black text-2xl">
          Gerenciador de Esferas do Drag√£o
        </h1>

        <BallFilterForm currentValue={selected} onSelect={onSelect} />
      </div>

      <div className="grid sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4">
        {balls.map((ball) => (
          <BallCard key={ball.id} ball={ball} profileId={1} />
        ))}
      </div>
    </>
  );
}
