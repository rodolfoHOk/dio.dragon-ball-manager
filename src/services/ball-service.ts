import { Ball } from '../models/ball';

export class BallService {
  private static balls: Ball[] = [
    {
      id: 1,
      name: '1 estrela',
      owner: 0,
      image: '/images/dragonball-1-star.png',
    },
    {
      id: 2,
      name: '2 estrelas',
      owner: 0,
      image: '/images/dragonball-2-stars.png',
    },
    {
      id: 3,
      name: '3 estrelas',
      owner: 0,
      image: '/images/dragonball-3-stars.png',
    },
    {
      id: 4,
      name: '4 estrelas',
      owner: 1,
      image: '/images/dragonball-4-stars.png',
    },
    {
      id: 5,
      name: '5 estrelas',
      owner: 0,
      image: '/images/dragonball-5-stars.png',
    },
    {
      id: 6,
      name: '6 estrelas',
      owner: 0,
      image: '/images/dragonball-6-stars.png',
    },
    {
      id: 7,
      name: '7 estrelas',
      owner: 0,
      image: '/images/dragonball-7-stars.png',
    },
  ];

  static getBalls(): Ball[] {
    return this.balls;
  }

  static foundedBall(ownerId: number, ballId: number) {
    this.balls = this.balls.map((ball) => {
      if (ball.id === ballId) {
        ball.owner = ownerId;
      }
      return ball;
    });
  }

  static hasAllBalls(ownerId: number) {
    const balls = this.balls.filter((ball) => ball.owner === ownerId);
    return balls.length === 7;
  }
}
