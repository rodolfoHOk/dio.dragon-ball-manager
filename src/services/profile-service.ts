import { Profile } from '../models/profile';

export class ProfileService {
  private static profiles: Profile[] = [
    {
      id: 1,
      name: 'Rudy HiOk',
      image:
        'https://raw.githubusercontent.com/rodolfoHOk/portfolio-img/main/images/avatar/pedro.jpg',
      balls: [1, 2, 4, 6, 7],
    },
    {
      id: 2,
      name: 'Jao Pones',
      image:
        'https://raw.githubusercontent.com/rodolfoHOk/portfolio-img/main/images/avatar/joao.jpg',
      balls: [3, 5],
    },
  ];

  static getProfile(id: number): Profile | undefined {
    return this.profiles.find((profile) => profile.id === id);
  }

  static addBall(profileId: number, ballId: number) {
    this.profiles = this.profiles.map((profile) => {
      if (profile.id === profileId) {
        profile.balls.push(ballId);
      } else {
        if (profile.balls.includes(ballId)) {
          profile.balls = profile.balls.filter((ball) => ball !== ballId);
        }
      }
      return profile;
    });
  }
}
