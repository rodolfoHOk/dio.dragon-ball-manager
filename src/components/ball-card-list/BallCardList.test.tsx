import * as Toast from '@radix-ui/react-toast';
import { fireEvent, render, screen } from '@testing-library/react';
import { BallCardList } from './BallCardList';
import { BallService } from '../../services/ball-service';

describe('ball card list component test', () => {
  it('should render ball card list', () => {
    const { container } = render(
      <BallCardList profileId={1} allBalls={BallService.getBalls()} />
    );

    expect(screen.getByTestId('filter-trigger')).toBeInTheDocument();
    expect(screen.getAllByTestId('card').length).toBe(7);
    const foundBadges = screen.getAllByText('Encontrada');
    const notFoundBadges = screen.getAllByText('Não encontrada');
    expect(foundBadges.length + notFoundBadges.length).toBe(7);
    expect(container).toMatchSnapshot();
  });

  it('should filter my balls', () => {
    const { container } = render(
      <BallCardList profileId={1} allBalls={BallService.getBalls()} />
    );

    const filterSelect = container.querySelector('select');
    if (filterSelect) {
      fireEvent.change(filterSelect, {
        target: { value: 'me' },
      });
    }

    expect(screen.getAllByTestId('card').length).toBe(5);
  });

  it('should filter not founded balls', () => {
    const { container } = render(
      <BallCardList profileId={1} allBalls={BallService.getBalls()} />
    );

    const filterSelect = container.querySelector('select');
    if (filterSelect) {
      fireEvent.change(filterSelect, {
        target: { value: 'not-me' },
      });
    }

    expect(screen.getAllByTestId('card').length).toBe(2);
  });

  it('should not filter when all is selected', () => {
    const { container } = render(
      <BallCardList profileId={1} allBalls={BallService.getBalls()} />
    );

    const filterSelect = container.querySelector('select');
    if (filterSelect) {
      fireEvent.change(filterSelect, {
        target: { value: 'all' },
      });
    }

    expect(screen.getAllByTestId('card').length).toBe(7);
  });

  it('should open validate ball modal', async () => {
    render(<BallCardList profileId={1} allBalls={BallService.getBalls()} />);

    const iFoundButton = screen.getByTestId('card-button-3');
    expect(iFoundButton).toBeInTheDocument();
    fireEvent.click(iFoundButton);

    const backButton = screen.getByText('Voltar');
    expect(backButton).toBeInTheDocument();
    fireEvent.click(backButton);
  });

  it('should show toast if enter with invalid code', async () => {
    render(
      <Toast.Provider>
        <div>
          <BallCardList profileId={1} allBalls={BallService.getBalls()} />
          <Toast.Viewport />
        </div>
      </Toast.Provider>
    );

    const iFoundButton = screen.getByTestId('card-button-3');
    expect(iFoundButton).toBeInTheDocument();
    fireEvent.click(iFoundButton);

    const validateButton = screen.getByTestId('validate-button-3');
    expect(validateButton).toBeInTheDocument();
    fireEvent.click(validateButton);

    expect(
      screen.getByText('Código da esfera do dragão não confere')
    ).toBeInTheDocument();

    const backButton = screen.getByText('Voltar');
    expect(backButton).toBeInTheDocument();
    fireEvent.click(backButton);

    expect(
      screen.queryByText('Código da esfera do dragão não confere')
    ).toBeNull();
  });

  it('should validate ball code', async () => {
    render(
      <Toast.Provider>
        <div>
          <BallCardList profileId={1} allBalls={BallService.getBalls()} />
          <Toast.Viewport />
        </div>
      </Toast.Provider>
    );

    let notFoundBadges = screen.getAllByText('Não encontrada');
    expect(notFoundBadges.length).toBe(2);
    let foundBadges = screen.getAllByText('Encontrada');
    expect(foundBadges.length).toBe(5);

    const iFoundButton = screen.getByTestId('card-button-3');
    expect(iFoundButton).toBeInTheDocument();
    fireEvent.click(iFoundButton);

    const codeInput = screen.getByTestId('ball-code-input');
    fireEvent.change(codeInput, { target: { value: 'esfera-3-valida' } });

    const validateButton = screen.getByTestId('validate-button-3');
    expect(validateButton).toBeInTheDocument();
    fireEvent.click(validateButton);

    expect(
      screen.queryByText('Código da esfera do dragão não confere')
    ).toBeNull();
    notFoundBadges = screen.getAllByText('Não encontrada');
    expect(notFoundBadges.length).toBe(1);
    foundBadges = screen.getAllByText('Encontrada');
    expect(foundBadges.length).toBe(6);
  });
});
