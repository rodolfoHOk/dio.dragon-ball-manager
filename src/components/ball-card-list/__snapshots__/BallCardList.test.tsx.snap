// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ball card list component test should render ball card list 1`] = `
<div>
  <div
    class="flex flex-col gap-4"
  >
    <div
      class="flex justify-between items-center"
    >
      <h2
        class="font-black text-2xl"
      >
        Gerenciador de Esferas do Drag達o
      </h2>
      <form
        class="flex flex-col gap-2"
      >
        <label
          class="font-bold text-lg"
        >
          Filtrar
        </label>
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r0:"
          aria-expanded="false"
          class="w-44 h-10 px-2 inline-flex items-center justify-between gap-1 bg-zinc-50 dark:bg-zinc-900 dark:hover:bg-zinc-800 hover:bg-zinc-100 border border-neutral-500 rounded shadow transition-colors duration-300"
          data-state="closed"
          data-testid="filter-trigger"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Todas as esferas
          </span>
          <span
            aria-hidden="true"
          >
            <svg
              class="lucide lucide-chevrons-up-down"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m7 15 5 5 5-5"
              />
              <path
                d="m7 9 5-5 5 5"
              />
            </svg>
          </span>
        </button>
        <select
          aria-hidden="true"
          style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          tabindex="-1"
        >
          <option
            selected=""
            value="all"
          >
            Todas as esferas
          </option>
          <option
            value="me"
          >
            Minhas esferas
          </option>
          <option
            value="not-me"
          >
            N達o tenho
          </option>
        </select>
      </form>
    </div>
    <div
      class="grid sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4"
    >
      <div
        class="w-full p-4 flex flex-col bg-white dark:bg-zinc-800 text-zinc-950 dark:text-zinc-50 rounded-lg border-2 border-neutral-300 transition-colors duration-300"
        data-testid="card"
      >
        <div>
          <img
            alt="1 estrela"
            src="/images/dragonball-1-star.png"
          />
        </div>
        <div
          class="flex flex-col items-center gap-6 mb-6"
        >
          <p
            class="font-bold text-lg"
          >
            1 estrela
          </p>
          <div
            class="w-full flex justify-center"
          >
            <div
              class="px-2 py-1 flex items-center justify-center rounded-md bg-green-600"
            >
              <span
                class="text-zinc-50 font-bold"
              >
                Encontrada
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex justify-center"
        />
      </div>
      <div
        class="w-full p-4 flex flex-col bg-white dark:bg-zinc-800 text-zinc-950 dark:text-zinc-50 rounded-lg border-2 border-neutral-300 transition-colors duration-300"
        data-testid="card"
      >
        <div>
          <img
            alt="2 estrelas"
            src="/images/dragonball-2-stars.png"
          />
        </div>
        <div
          class="flex flex-col items-center gap-6 mb-6"
        >
          <p
            class="font-bold text-lg"
          >
            2 estrelas
          </p>
          <div
            class="w-full flex justify-center"
          >
            <div
              class="px-2 py-1 flex items-center justify-center rounded-md bg-green-600"
            >
              <span
                class="text-zinc-50 font-bold"
              >
                Encontrada
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex justify-center"
        />
      </div>
      <div
        class="w-full p-4 flex flex-col bg-white dark:bg-zinc-800 text-zinc-950 dark:text-zinc-50 rounded-lg border-2 border-neutral-300 transition-colors duration-300"
        data-testid="card"
      >
        <div>
          <img
            alt="3 estrelas"
            src="/images/dragonball-3-stars.png"
          />
        </div>
        <div
          class="flex flex-col items-center gap-6 mb-6"
        >
          <p
            class="font-bold text-lg"
          >
            3 estrelas
          </p>
          <div
            class="w-full flex justify-center"
          >
            <div
              class="px-2 py-1 flex items-center justify-center rounded-md bg-red-600"
            >
              <span
                class="text-zinc-50 font-bold"
              >
                N達o encontrada
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex justify-center"
        >
          <button
            aria-controls="radix-:rb:"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="h-10 px-4 flex flex-row items-center justify-center font-bold rounded  transition-colors duration-300 bg-orange-400 text-red-800 hover:bg-orange-300 w-full"
            data-state="closed"
            data-testid="card-button-3"
            type="button"
          >
            Encontrei
          </button>
        </div>
      </div>
      <div
        class="w-full p-4 flex flex-col bg-white dark:bg-zinc-800 text-zinc-950 dark:text-zinc-50 rounded-lg border-2 border-neutral-300 transition-colors duration-300"
        data-testid="card"
      >
        <div>
          <img
            alt="4 estrelas"
            src="/images/dragonball-4-stars.png"
          />
        </div>
        <div
          class="flex flex-col items-center gap-6 mb-6"
        >
          <p
            class="font-bold text-lg"
          >
            4 estrelas
          </p>
          <div
            class="w-full flex justify-center"
          >
            <div
              class="px-2 py-1 flex items-center justify-center rounded-md bg-green-600"
            >
              <span
                class="text-zinc-50 font-bold"
              >
                Encontrada
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex justify-center"
        />
      </div>
      <div
        class="w-full p-4 flex flex-col bg-white dark:bg-zinc-800 text-zinc-950 dark:text-zinc-50 rounded-lg border-2 border-neutral-300 transition-colors duration-300"
        data-testid="card"
      >
        <div>
          <img
            alt="5 estrelas"
            src="/images/dragonball-5-stars.png"
          />
        </div>
        <div
          class="flex flex-col items-center gap-6 mb-6"
        >
          <p
            class="font-bold text-lg"
          >
            5 estrelas
          </p>
          <div
            class="w-full flex justify-center"
          >
            <div
              class="px-2 py-1 flex items-center justify-center rounded-md bg-red-600"
            >
              <span
                class="text-zinc-50 font-bold"
              >
                N達o encontrada
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex justify-center"
        >
          <button
            aria-controls="radix-:rh:"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="h-10 px-4 flex flex-row items-center justify-center font-bold rounded  transition-colors duration-300 bg-orange-400 text-red-800 hover:bg-orange-300 w-full"
            data-state="closed"
            data-testid="card-button-5"
            type="button"
          >
            Encontrei
          </button>
        </div>
      </div>
      <div
        class="w-full p-4 flex flex-col bg-white dark:bg-zinc-800 text-zinc-950 dark:text-zinc-50 rounded-lg border-2 border-neutral-300 transition-colors duration-300"
        data-testid="card"
      >
        <div>
          <img
            alt="6 estrelas"
            src="/images/dragonball-6-stars.png"
          />
        </div>
        <div
          class="flex flex-col items-center gap-6 mb-6"
        >
          <p
            class="font-bold text-lg"
          >
            6 estrelas
          </p>
          <div
            class="w-full flex justify-center"
          >
            <div
              class="px-2 py-1 flex items-center justify-center rounded-md bg-green-600"
            >
              <span
                class="text-zinc-50 font-bold"
              >
                Encontrada
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex justify-center"
        />
      </div>
      <div
        class="w-full p-4 flex flex-col bg-white dark:bg-zinc-800 text-zinc-950 dark:text-zinc-50 rounded-lg border-2 border-neutral-300 transition-colors duration-300"
        data-testid="card"
      >
        <div>
          <img
            alt="7 estrelas"
            src="/images/dragonball-7-stars.png"
          />
        </div>
        <div
          class="flex flex-col items-center gap-6 mb-6"
        >
          <p
            class="font-bold text-lg"
          >
            7 estrelas
          </p>
          <div
            class="w-full flex justify-center"
          >
            <div
              class="px-2 py-1 flex items-center justify-center rounded-md bg-green-600"
            >
              <span
                class="text-zinc-50 font-bold"
              >
                Encontrada
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex justify-center"
        />
      </div>
    </div>
  </div>
</div>
`;
